### PR TITLE
Fill in all missing `upstream_repository`

### PR DESCRIPTION
**Pull request by Avasam**
I was navigating https://alexwaygood.github.io/typeshed-stats/stats-csv/ for unrelated reasons and noticed a lot of missing upstream source / url, and decided it'd be fun to go on a hunt.

The only actually closed-source project seems to be SAP's hdbcli.
Most of these were linked in their docs (although sometimes very obscure). Except for python-ini which has no documentation and no link on PyPI. But it's the same author so I assume it's the right repository.

I've updated the `upstream_repository` field validation to make the GDB change pass. I believe this respects the original intent of avoiding extra query params and fragments that may have accidentally been copied. And added more detailed error messages.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
(It's possible that the validation we're doing in `parse_metadata.py` is overly strict at the moment)

File: stubs/pyaudio/METADATA.toml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 version = "0.2.*"
-# There is no web portal for the source, this is the official source link:
-# https://people.csail.mit.edu/hubert/pyaudio/#sources
-upstream_repository = "https://people.csail.mit.edu/hubert/git/pyaudio.git"
+# There is no web portal for the source, this is the official source link
+upstream_repository = "https://people.csail.mit.edu/hubert/pyaudio/#sources"
```
Comment (Avasam) [inline stubs/pyaudio/METADATA.toml:3]: @AlexWaygood This is my usage of a git URL or a fragment. I'll let you decide on whether this value or the restrictive parsing should change, depending on your needs. It's definitely an edge case.

File: tests/parse_metadata.py [hunk 2]
```diff
@@ -203,9 +205,13 @@ def read_metadata(distribution: str) -> StubMetadata:
...
         parsed_url = urllib.parse.urlsplit(upstream_repository)
         assert parsed_url.scheme == "https", "URLs in the upstream_repository field should use https"
-        assert not parsed_url.netloc.startswith("www."), "`www.` should be removed from URLs in the upstream_repository field"
-        assert not parsed_url.query
-        assert not parsed_url.fragment
+        assert not parsed_url.netloc.startswith(
+            "www."
+        ), "`World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
+        assert (
+            parsed_url.hostname not in QUERY_URL_ALLOWLIST and not parsed_url.query
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:212]:
```suggestion
            parsed_url.hostname in QUERY_URL_ALLOWLIST or (not parsed_url.query)
```

File: stubs/pyaudio/METADATA.toml (hunk 1 above)
Comment (Avasam) [reply stubs/pyaudio/METADATA.toml]: I've updated the `upstream_repository` field validation. I believe this respects the original intent of avoiding extra query params and fragments that may have accidentally been copied.

File: stubs/beautifulsoup4/METADATA.toml [hunk 3]
```diff
@@ -1,4 +1,5 @@
 version = "4.12.*"
+upstream_repository = "https://git.launchpad.net/beautifulsoup"
```
Comment (AlexWaygood) [inline stubs/beautifulsoup4/METADATA.toml:2]: Maybe this would be more useful? I generally don't care too much about the git log when I'm looking at an upstream repository to review a typeshed PR, I just wanna see the source 😆

```suggestion
upstream_repository = "https://git.launchpad.net/beautifulsoup/tree"
```

File: tests/parse_metadata.py [hunk 4]
```diff
@@ -31,6 +31,8 @@
 
 
 _STUBTEST_PLATFORM_MAPPING: Final = {"linux": "apt_dependencies", "darwin": "brew_dependencies", "win32": "choco_dependencies"}
+QUERY_URL_ALLOWLIST = {"sourceware.org"}
+"""Some older websites have a bad pattern of using query params for navigation"""
```
Comment (Avasam) [inline tests/parse_metadata.py:35]: Feel free to wordsmith this.

File: stubs/gdb/METADATA.toml [hunk 5]
```diff
@@ -1,4 +1,7 @@
 version = "12.1.*"
+# This is the official web portal for GDB,
+# see https://sourceware.org/gdb/current/ for other ways of obtaining the source code.
+upstream_repository = "https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git"
```
Comment (AlexWaygood) [inline stubs/gdb/METADATA.toml:4]: Idk how many people are really going to be interested in the source code for `gdb` (or if this will pass our validation checks), but:

```suggestion
upstream_repository = "https://sourceware.org/git/?p=binutils-gdb.git;a=tree"
```

File: stubs/pyaudio/METADATA.toml [hunk 6]
```diff
@@ -1,4 +1,7 @@
 version = "0.2.*"
+# There is no web portal for the source, this is the official source link:
+# https://people.csail.mit.edu/hubert/pyaudio/#sources
+upstream_repository = "https://people.csail.mit.edu/hubert/git/pyaudio.git"
```
Comment (AlexWaygood) [inline stubs/pyaudio/METADATA.toml:4]: This is a bit weird; neither link would take us to a version of the source that you could view in a web browser. I think I'd prefer just doing this tbh:

```suggestion
# There is no web portal for the source, this is the official source link:
# https://people.csail.mit.edu/hubert/pyaudio/#sources
```

File: tests/parse_metadata.py (hunk 4 above)
Comment (AlexWaygood) [inline tests/parse_metadata.py:35]:
```suggestion
# Some older websites have a bad pattern of using query params for navigation
QUERY_URL_ALLOWLIST = {"sourceware.org"}
```

File: tests/parse_metadata.py [hunk 7]
```diff
@@ -203,9 +205,13 @@ def read_metadata(distribution: str) -> StubMetadata:
     if isinstance(upstream_repository, str):
         parsed_url = urllib.parse.urlsplit(upstream_repository)
         assert parsed_url.scheme == "https", "URLs in the upstream_repository field should use https"
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:207]:
```suggestion
        assert parsed_url.scheme == "https", f"{distribution}: URLs in the upstream_repository field should use https"
```

File: tests/parse_metadata.py [hunk 8]
```diff
@@ -203,9 +205,13 @@ def read_metadata(distribution: str) -> StubMetadata:
...
-        assert not parsed_url.netloc.startswith("www."), "`www.` should be removed from URLs in the upstream_repository field"
-        assert not parsed_url.query
-        assert not parsed_url.fragment
+        assert not parsed_url.netloc.startswith(
+            "www."
+        ), "`World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:210]:
```suggestion
        ), f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
```

File: tests/parse_metadata.py [hunk 9]
```diff
@@ -203,9 +205,13 @@ def read_metadata(distribution: str) -> StubMetadata:
...
+        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (
+            not parsed_url.query
+        ), "Query params (`?`) should be removed from URLs in the upstream_repository field"
+        assert not parsed_url.fragment, "Fragments (`#`) should be removed from URLs in the upstream_repository field"
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:214]:
```suggestion
        ), f"{distribution}: Query params (`?`) should be removed from URLs in the upstream_repository field"
        assert not parsed_url.fragment, f"{distribution}: Fragments (`#`) should be removed from URLs in the upstream_repository field"
```

**Review by AlexWaygood — COMMENTED** [review]
Thanks!

File: stubs/docutils/METADATA.toml [hunk 10]
```diff
@@ -1,4 +1,5 @@
 version = "0.20.*"
+upstream_repository = "https://sourceforge.net/p/docutils/code"
```
Comment (AlexWaygood) [inline stubs/docutils/METADATA.toml:2]: Oh, could you also add sourceforge to this set here?

https://github.com/python/typeshed/blob/df08fcec5f281ba01266869ec55368819aed0a77/scripts/create_baseline_stubs.py#L112

File: stubs/beautifulsoup4/METADATA.toml (hunk 3 above)
Comment (Avasam) [reply stubs/beautifulsoup4/METADATA.toml:2]: Agreed. For example docutils' sourceforge link goes directly to the code tab.
I'll update the launchpad link in pytz comment too

File: stubs/gdb/METADATA.toml (hunk 5 above)
Comment (Avasam) [reply stubs/gdb/METADATA.toml:4]: It should, it's just part of the queryparam, will try

Comment (AlexWaygood) [reply stubs/gdb/METADATA.toml:4]: is it obvious i'm not a web developer

File: tests/parse_metadata.py (hunk 4 above)
Comment (Avasam) [reply tests/parse_metadata.py:35]: I don't mind either way, but I liked the docstring showing up in IDEs ^^
image

Comment (AlexWaygood) [reply tests/parse_metadata.py:35]: To me it just looks pretty odd doing something I thought was Sphinx-specific when we don't autogenerate any docs from this file. I'd much prefer the comment here, I'm afraid :)

File: stubs/docutils/METADATA.toml (hunk 10 above)
Comment (Avasam) [reply stubs/docutils/METADATA.toml:2]: I can. It seems the logic will truncate to `https://sourceforge.net/p/repo`, is that good here? I could special-case sourceforge to add back `/code`

Comment (AlexWaygood) [reply stubs/docutils/METADATA.toml:2]: Okay, since this is non-trivial in complexity, let's postpone it to a followup PR :)

File: tests/parse_metadata.py [hunk 11]
```diff
@@ -202,10 +204,16 @@ def read_metadata(distribution: str) -> StubMetadata:
...
+        assert parsed_url.scheme == "https", f"{distribution}: URLs in the upstream_repository field should use https"
+        assert not parsed_url.netloc.startswith(
+            "www."
+        ), f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
+        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (
+            not parsed_url.query
+        ), f"{distribution}: Query params (`?`) should be removed from URLs in the upstream_repository field"
+        assert (
+            not parsed_url.fragment
+        ), f"{distribution}: Fragments (`#`) should be removed from URLs in the upstream_repository field"
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:216]: I can't bear the way black formats these lines, I'm afraid ;)

```suggestion
        no_www_please = (
            f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
        )
        assert not parsed_url.netloc.startswith("www."), no_www_please
        no_query_params_please = (
            f"{distribution}: Query params (`?`) should be removed from URLs in the upstream_repository field"
        )
        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (not parsed_url.query), no_query_params_please
        no_fragments_please = f"{distribution}: Fragments (`#`) should be removed from URLs in the upstream_repository field"
        assert not parsed_url.fragment, no_fragments_please
```

File: tests/parse_metadata.py [hunk 12]
```diff
@@ -31,6 +31,8 @@
 
 
 _STUBTEST_PLATFORM_MAPPING: Final = {"linux": "apt_dependencies", "darwin": "brew_dependencies", "win32": "choco_dependencies"}
+# Some older websites have a bad pattern of using query params for navigation.
+QUERY_URL_ALLOWLIST = {"sourceware.org"}
```
Comment (AlexWaygood) [inline tests/parse_metadata.py:35]: Actually, yes, could you make this private? Not a big deal, but it would be consistent with `_STUBTEST_PLATFORM_MAPPING` immediately above

File: tests/parse_metadata.py (hunk 11 above)
Comment (Avasam) [reply tests/parse_metadata.py:216]: This is one of my big pet peaves with black too. I'll go with your changes.

File: tests/parse_metadata.py (hunk 12 above)
Comment (Avasam) [reply tests/parse_metadata.py:35]:
```suggestion
_QUERY_URL_ALLOWLIST = {"sourceware.org"}
```

File: tests/parse_metadata.py [hunk 13]
```diff
@@ -202,10 +204,16 @@ def read_metadata(distribution: str) -> StubMetadata:
...
         parsed_url = urllib.parse.urlsplit(upstream_repository)
-        assert parsed_url.scheme == "https", "URLs in the upstream_repository field should use https"
-        assert not parsed_url.netloc.startswith("www."), "`www.` should be removed from URLs in the upstream_repository field"
-        assert not parsed_url.query
-        assert not parsed_url.fragment
+        assert parsed_url.scheme == "https", f"{distribution}: URLs in the upstream_repository field should use https"
+        assert not parsed_url.netloc.startswith(
+            "www."
+        ), f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
+        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (
```
Comment (Avasam) [inline tests/parse_metadata.py:211]:
```suggestion
        assert parsed_url.hostname in _QUERY_URL_ALLOWLIST or (
```

File: tests/parse_metadata.py [hunk 14]
```diff
@@ -202,10 +204,17 @@ def read_metadata(distribution: str) -> StubMetadata:
...
-        assert not parsed_url.fragment
+        assert parsed_url.scheme == "https", f"{distribution}: URLs in the upstream_repository field should use https"
+        no_www_please = (
+            f"{distribution}: `World Wide Web` subdomain (`www.`) should be removed from URLs in the upstream_repository field"
+        )
+        assert not parsed_url.netloc.startswith("www."), no_www_please
+        no_query_params_please = (
+            f"{distribution}: Query params (`?`) should be removed from URLs in the upstream_repository field"
+        )
+        assert parsed_url.hostname in QUERY_URL_ALLOWLIST or (not parsed_url.query), no_query_params_please
```
Comment (Avasam) [inline tests/parse_metadata.py:215]:
```suggestion
        assert parsed_url.hostname in _QUERY_URL_ALLOWLIST or (not parsed_url.query), no_query_params_please
```

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Comment by srittau** [comment]
Thanks! These are actually the packages where the `upstream_repository` field is most useful and why I wanted it in the first place.
